### ISSUE TITLE
Support private servers

### ISSUE THREAD

**Issue by IvanSanchez** [Rich-Harris/degit#19]
Splitting off from the conversation at #16:

@melniker says:

>I was planning on working on non-github support [...], and my plan included a config json file that gave the characteristics of the servers. I was going to pre-populate it for github/gitlab/bitbucket but it could then be easily extendable for other servers. It would specify every property you'd need, including URI format & location of the tarball.
>
> This way, degit would be easily extendable not only for public git servers (which could be put into debit/master via PR) but also for private instances inside a big corporation, e.g.
>
>Could also of course take these config settings via cli parameters.
>
>Thoughts?

I, on the other hand, would prefer something of the form:

```
gitlab-https://my-own-gitlab-server.local/my-user/-my-repo.git
```

Maybe also parsing the repo URL and checking if the substrings `gitlab` or `bitbucket` (and so on) are in there. Personally I'd prefer as little configuration as possible.

**Comment by Rich-Harris** [comment]
marking this and #20 as 'pr welcome'! Am in favour in principle, but it's not something I personally have a use for. I think there might be some config required, just to figure out what the correct URL for the `.tar.gz` file is? It seems to differ from service to service.

**Comment by melniker** [comment]
@IvanSanchez, I may be a bit config heavy :)
But, with so many different implementations listed in #20, I think at least the archive file locations (as @Rich-Harris mentions) should be listed in a config file, outside of index.js. 

Just my $.02; I don't have much time in the near term to help work on this :/

**Comment by Rich-Harris** [comment]
Closed via #29
